### PR TITLE
feat: query for info only from ticks that satisfy specific prop states conditions

### PR DESCRIPTION
**Pull request by dankotov**
bar the translation of user input into PropInfo vector

### CODE REVIEW

**Comment by LaihoE** [comment]
looks good so far, check discord for some more info

**Comment by dankotov** [comment]
@LaihoE tested to work with basic condition like 

```
let wanted_props = vec!["X".to_string()];
let mut wanted_prop_states: AHashMap<String, Variant> = AHashMap::new();
wanted_prop_states.insert(
    "CCSGameRulesProxy.CCSGameRules.m_bBombPlanted".to_string(),
    Variant::Bool(false),
);

let huf = create_huffman_lookup_table();
let settings = ParserInputs {
    wanted_players: vec![],
    real_name_to_og_name: AHashMap::default(),
    wanted_player_props: wanted_props.clone(),
    wanted_events: vec![],
    wanted_other_props: vec!["CCSGameRulesProxy.CCSGameRules.m_bBombPlanted".to_string()],
    parse_ents: true,
    wanted_ticks: vec![],
    parse_projectiles: true,
    only_header: false,
    count_props: false,
    only_convars: false,
    huffman_lookup_table: &huf,
    order_by_steamid: false,
    wanted_prop_states: wanted_prop_states,
};
let mut ds = Parser::new(settings, ParsingMode::Normal);
let file = File::open(path_to_demo).unwrap();
let mmap = unsafe { MmapOptions::new().map(&file).unwrap() };
let _output = ds.parse_demo(&mmap).unwrap();
```

File: src/parser/src/first_pass/prop_controller.rs [hunk 1]
```diff
@@ -119,6 +132,19 @@ impl PropController {
...
+                    base: PropInfo {
+                        id: someid2,
+                        prop_type: PropType::Button,
+                        prop_name: bn.to_string(),
+                        prop_friendly_name: bn.to_string(),
+                        is_player_prop: true,
+                    },
+                    wanted_prop_state: wanted_state.clone(),
+                });
+                someid2 += 1;
```
Comment (dankotov) [inline src/parser/src/first_pass/prop_controller.rs:146]: should i be creating a new ID (`someid2`) here or reusing `someid` is fine?

File: src/parser/src/first_pass/prop_controller.rs [hunk 2]
```diff
@@ -132,6 +158,18 @@ impl PropController {
                 is_player_prop: true,
             });
         }
+        if let Some(wanted_state) = self.wanted_prop_states.get(&("active_weapon_original_owner".to_string())) {
```
Comment (dankotov) [inline src/parser/src/first_pass/prop_controller.rs:161]: I don't like how all of this code works with a bunch of if statements. Is there a reason all of the custom props are not stored as a map and iterated over keys conditionally (if the prop is  in `wanted_player_props`) pushing each prop's metadata (value in the map)  into `prop_infos`?

I can refactor if the current approach is just a draft.

File: src/parser/src/first_pass/prop_controller.rs [hunk 3]
```diff
@@ -423,6 +749,22 @@ impl PropController {
...
+                    base: PropInfo {
+                        id: f.prop_id as u32,
+                        prop_type: *prop_type,
+                        prop_name: prop_name.to_string(),
+                        prop_friendly_name: self
+                            .real_name_to_og_name
+                            .get(&prop_name.to_string())
+                            .unwrap_or(&(prop_name.to_string()))
+                            .to_string(),
+                        is_player_prop: false, // does this actually affect anything?
```
Comment (dankotov) [inline src/parser/src/first_pass/prop_controller.rs:763]: does it matter whether prop is a player prop or not?

File: src/parser/src/first_pass/prop_controller.rs (hunk 1 above)
Comment (LaihoE) [reply src/parser/src/first_pass/prop_controller.rs:146]: You should be using the same one for it to work.

**Comment by LaihoE** [comment]
Sorry for the late reply. 

Yes prop_controller is some horrible code that needs to be refactored. And very much true that the if statements were already getting quite ridiculous, and now double the amount is maybe a little too much. If you want to try to simplify it then that would be great.

As for "is_player_prop" it is used in game events for something. I think it is to know where to look for the value (from the perspective of the player or something i don't quite remember) but maybe this could also be implemented in some other way.

File: src/parser/src/first_pass/prop_controller.rs (hunk 1 above)
Comment (dankotov) [reply src/parser/src/first_pass/prop_controller.rs:146]: Would it be ok then if the id is incremented twice for the same button? In the case if that button is in both - `wanted_player_props` and `wanted_prop_state_infos`? Or incremented once, but in the case that it was not in `wanted_player_props` and was in `wanted_prop_state_infos`?

**Comment by dankotov** [comment]
@LaihoE no worries on the late reply

Yeah, I can refactor those if statements. I just have a few more questions to better understand the `set_custom_propinfos` function.

If I interpret it correctly, sets PropInfo for A) Buttons B) Ticks, SteamID, Name C) Props that are of type `PropType::Custom`. The A and B part I think are fine, the wealth of the code is caused by C. I see that there is a `TYPEHM` map in `maps.rs` which maps prop string name to `PropType`. My initial idea was to retrieve all key from that map, for which the value is `PropType::Custom`, iterate over the keys. However, I noticed that not all prop names referenced in the if statements were present in `TYPEHM`. Is that by design? Or were they just missed accidentally? If I could add the missing prop names to `TYPEHM` I think this refactor would be elegant and quick.

As for the `is_player_prop`, at the moment, I believe, whether a prop is a player prop is derived by which parser input array it was put into by the end user - `wanted_player_props` or `wanted_other_props`. Is there a reason why we dont hardcode props (using some kind of a map, perhaps just a name to boolean) to whether they are player props or not?

**Comment by LaihoE** [comment]
You can just add all the A) Buttons B) Ticks, SteamID, Name C) as PropType::Custom :+1:

You are probably right about the is_player_prop. I don't really remember much about its role.

**Comment by LaihoE** [comment]
Are we ready to merge?

**Comment by dankotov** [comment]
Not yet. Need to add 1 more Rust test case and JS bindings.

**Comment by dankotov** [comment]
^ Had to remove other wanted states as querying for prop states while parsing event(s) not yet implemented and "other" props do not work when parsing ticks.

Will probably implement querying for prop states while parsing event(s) in another PR down the line.

**Comment by LaihoE** [comment]
LGTM. Will create a new release soon.

**Pull request merged**